lib/store.ml

```diff
@@ -76,16 +97,41 @@ let merge (s1 : t) (s2 : t) : t =
   let len2 = Array.length s2 in
   let nlen = if len1 < len2 then len2 else len1 in
   let ans = allocate nlen in
-  (** Important: No more allocation from here. *)
-  Array.blit s2 0 ans 0 len2;
-  for i = 0 to pred len1 do
-    let v = Array.unsafe_get s1 i in
-    match v with
-    | None -> ()
-    | Some _ -> Array.unsafe_set ans i v
+  let merge i x y =
+    let fn = Array.unsafe_get !merge_fns i in
+    let fn : 'a merge_field = Obj.magic fn in
+    match x, y with
+    | Some x, Some y -> fn (Both (x, y))
+    | Some x, None -> fn (Left x)
+    | None, Some y -> fn (Right y)
+    | None, None -> None
+  in
+  for i = 0 to pred nlen do
+    let v1 = if i < len1 then Array.unsafe_get s1 i else None in
+    let v2 = if i < len2 then Array.unsafe_get s2 i else None in
+    Array.unsafe_set ans i (merge i v1 v2)
   done;
   ans
 
-let field () = next ()
+let field merge =
+  let i = next () in
+  let () =
+    (** Extend merge array if too small *)
+    let init = !merge_fns in
+    let len = Array.length init in
+    if i < len then ()
+    else begin
+      let arr = allocate (len + 1) in
```

Comment:
It's not critical as there will probably be few of these but: `allocate (len * 2)` would be better.